pkg/system/autoscaler.go

```diff
@@ -143,11 +152,88 @@ func (r *Reconciler) ensureHPAV1Cleanup(log *logrus.Entry) error {
 }
 
 func (r *Reconciler) ensureHPAV2Cleanup(log *logrus.Entry) error {
-	log.Infof("HPAV2 implementation pending")
+
+	noobaaHpaSelector, _ := labels.Parse("app=noobaa")
+	noobaaHPAV2FieldSelector, _ := fields.ParseSelector("metadata.name=hpav2-noobaa-endpoint")
+	// List autoscalers based on the label and name for HPAV2 ,
+	// Added name beause autoscalingv2 KubeList will fetch autoscaler with version v1 and v2
+	autoscalersv2 := &autoscalingv2.HorizontalPodAutoscalerList{}
+	util.KubeList(autoscalersv2, &client.ListOptions{Namespace: r.Request.Namespace, LabelSelector: noobaaHpaSelector,
+		FieldSelector: noobaaHPAV2FieldSelector})
+
+	if len(autoscalersv2.Items) == 0 {
+		return nil
+	}
+	log.Infof("Delete HPAV2 autoscaler with name %s", autoscalersv2.Items[0].Name)
+	// Dele HPAV2 resourses, role rolebinding, serviceAccount, service, deployments, APIService etc
+	deleteHPAV2Resources(r.Request.Namespace)
+	if !util.KubeDelete(&autoscalersv2.Items[0]) {
+		log.Errorf("Falied to delete HPAV2 existing %q autoscaler", autoscalersv2.Items[0].Name)
+		return fmt.Errorf("Falied to delete HPAV2 existing %q autoscaler", autoscalersv2.Items[0].Name)
+	}
+	autoscalersv2 = &autoscalingv2.HorizontalPodAutoscalerList{}
+	util.KubeList(autoscalersv2, &client.ListOptions{Namespace: r.Request.Namespace, LabelSelector: noobaaHpaSelector,
+		FieldSelector: noobaaHPAV2FieldSelector})
+	if len(autoscalersv2.Items) > 0 {
+		// Return error if HPAV2 autoscaler not deleted
+		return fmt.Errorf("HPAV2 autoscaler still not deleted")
+	}
 
 	return nil
 }
 
+func deleteHPAV2Resources(namespace string) {
+	log := util.Logger().WithField("func", "deleteHPAV2Resources")
+	hpav2Selector, _ := labels.Parse("app=prometheus-adapter")
+	clusterRoleBindings := &rbacv1.ClusterRoleBindingList{}
+	clusterRoles := &rbacv1.ClusterRoleList{}
+	rolesBindings := &rbacv1.RoleBindingList{}
+	roles := &rbacv1.RoleList{}
+
+	deployments := &appsv1.DeploymentList{}
+	services := &corev1.ServiceList{}
+	apiServices := &apiregistration.APIServiceList{}
+
+	util.KubeList(clusterRoleBindings, &client.ListOptions{LabelSelector: hpav2Selector})
+	util.KubeList(clusterRoles, &client.ListOptions{LabelSelector: hpav2Selector})
+	util.KubeList(rolesBindings, &client.ListOptions{LabelSelector: hpav2Selector})
+	util.KubeList(roles, &client.ListOptions{Namespace: namespace, LabelSelector: hpav2Selector})
+
+	util.KubeList(deployments, &client.ListOptions{Namespace: namespace, LabelSelector: hpav2Selector})
+	util.KubeList(services, &client.ListOptions{Namespace: namespace, LabelSelector: hpav2Selector})
+	util.KubeList(apiServices, &client.ListOptions{Namespace: namespace, LabelSelector: hpav2Selector})
+
+	for _, obj := range clusterRoleBindings.Items {
+		log.Warnf("clusterRoleBindings %q removing without grace", obj.Name)
+		util.KubeDelete(&obj, client.GracePeriodSeconds(0))
```

Comment:
It might be worth considering examining the return value of `util.KubeDelete` in this and subsequent cases, and `deleteHPAV2Resources()`  could either handle or return an error.